feat: create pull request template

eddybrando opened the pull request:
## Changes

Created a pull request template.
Closes #2455

## Implementation details

The file name and location follows the guide in https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository.

## Next steps if this PR gets merged

- Create similar PR templates for the other projects' repositories.
- Update the "Pull requests" chapter in the **CONTRIBUTING.md** file.
- Create a discussion to decide whether to add a "PR Checklist" section and how if should be.

Tansito commented:
This is really nice @eddybrando, should we update too the **CONTRIBUTING.md** in case this is merged (or in the same PR)?

abdonrd commented:
I like the idea of having a pull request template! 👏🏻 

But I would also like it to be much more concise. All the PR checklist seems heavy to me. As a barrier to entry to contribute. If we want to be strict with this type of good practices, I think it would be enough to have them defined in the contributing, and let the reviewers take care that they have been fulfilled before approving. And so the description of the PR is cleaner.

Also personally I don't like to use the checkboxes for this, because they appear like tasks:

<img width="566" alt="Screenshot 2022-01-26 at 11 52 44" src="https://user-images.githubusercontent.com/1007051/151150650-dbcde56c-55ef-418f-b5ad-69f32039582d.png">

I imagine something like the reference to the issue that it's solving, an explanation of why you have solved it in that particular way, maybe a before and after (with image/video would be fantastic), ...

techtolentino commented:
@eddybrando - thanks for putting together the PR template. I'll say that I've been on teams (much larger, and not open-source project) where we did have a checkbox list of criteria, and it became a pattern for people to casually check off items, without _really_ performing that check, ultimately defeating the purpose :/ 

What do you think of just having a bulleted list of items, sort of like, friendly reminders?


Checklist:
* Did you follow the [contributing guidelines](https://github.com/Qiskit/qiskit.org/blob/main/CONTRIBUTING.md)
* Be sure to test your code, otherwise open a [new issue]() to handle testing
* Please update any relevant documentation that is needed


Not sure if that'd lead to unwanted noise in the PRs. just a thought...

abdonrd commented:
As another option for the reminders, we also can use a hidden top comment in the PR. Example:

<img width="1237" alt="Screenshot 2022-01-26 at 18 21 35" src="https://user-images.githubusercontent.com/1007051/151214010-459464eb-a23b-46d6-8173-5bf0181f2541.png">

And GitHub UI already has the `Helpful resources` on the right and the information message at the bottom with:
> Remember, contributions to this repository should follow its contributing guidelines and code of conduct.

eddybrando commented:
Thanks a lot @Tansito for your suggestion! I added a "next steps" section on the description of this PR including the **CONTRIBUTING.md** update.

---

Thanks a lot @abdonrd and @techtolentino for your input.

I like @abdonrd's suggestion about adding a "why". I think the "why" can be part of the "implementation details" section, so I extended the description to make it clearer.

I also extended the "screenshots" section similarly to suggest adding a "before" and "after" situation if possible and useful. I preferred this over creating two new sections because not all changes are visual and that might make the unnecessary part for a lot of the PRs longer.

In regards to the "checklist" section, I really appreciate both of your inputs, and like @techtolentino's idea of converting the checklist into a regular list particularly. But here I believe we can have a more in depth discussion to try and find the correct balance, so I added as a next step the creation of a GitHub discussion about this in case this PR gets merged.

---

I'm happy to hear your input and maybe get this merged as a first step towards improving our code review process. 🙂

abdonrd commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,51 @@
+  Example:
+  Implemented the functionality to update the user's name.
+  Closes #42
+  Co-authored-by: @octocat
```
This works outside of the commit description?

https://github.blog/2018-01-29-commit-together-with-co-authors/

eddybrando replied:
I guess that if we "squash and merge" the PR, then it ends up in the commit description, and that's our usual merging strategy, so it should work as intended.

abdonrd commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,51 @@
+  and add co-authors where appropriate with the "co-authored-by" keyword.
+
+  Example:
+  Implemented the functionality to update the user's name.
```
This information could be directly the title?

eddybrando replied:
If it is very short then yes, since we're mostly using "squash and merge" for PRs and the title ends up being the commit message.

But here is a good place to be a little bit more descriptive to help the code reviews.

abdonrd commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,51 @@
+  Please start reviewing the changes to the controller files. Then check the
+  changes to the database files. This way, you will have more context about the
+  changes made to the database model.
+  You can test the changes at https://preview-42.example.com/profile/update
```
We have this information directly in the GitHub UI:

<img width="885" alt="Screenshot 2022-01-27 at 10 08 51" src="https://user-images.githubusercontent.com/1007051/151327748-88a203a1-30d5-4bd4-bccd-8f92de5bf91d.png">
<img width="868" alt="Screenshot 2022-01-27 at 10 09 07" src="https://user-images.githubusercontent.com/1007051/151327754-d2e4cd0d-d330-4355-8bef-6551814084dd.png">

eddybrando replied:
Yes, but that action displays the index of the preview PR.

Here we have also the opportunity to point to an specific page to quickly direct the code reviewer to the place they need to see without having to first navigate to it by themselves.

abdonrd commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,51 @@
+  are not in the scope of this PR.
+-->
+
+## How to read this PR
+
+<!--
+  Optional.
+
+  If useful for the code review, add some guidance for how to read this PR,
+  including links to preview builds.
```
Maybe we can merge this section with the `Implementation details` section? What do you think? To have less sections.

eddybrando replied:
Maaaaybe 🤔... I like the part of having less sections, but I also think that they are different enough to have their own to be honest.

abdonrd reviewed (approved):
I personally like templates with fewer sections, otherwise I feel like it overwhelms me.

But I think we can move on! Thanks Eddy!

And I'm still open to the thoughts of the other team members.

techtolentino reviewed (approved):
I think this is a good balance, and captures what we should expect from a PR. Like many templates (issue or PR), I think after a few uses, we'll get used to it and it'll become second nature.

For new contributors, however, this sets them up with a good amount of guidance/detail.

Thank you @eddybrando 

🚢

eddybrando commented:
Thanks a lot everyone for your contributions to this PR!

I'll merge this now and start with the next steps detailed in the PR description.
